### PR TITLE
[SPARK-20650][core] Remove JobProgressListener.

### PR DESCRIPTION
**Pull request by vanzin**
The only remaining use of this class was the SparkStatusTracker, which
was modified to use the new status store. The test code to wait for
executors was moved to TestUtils and now uses the SparkStatusTracker API.

Indirectly, ConsoleProgressBar also uses this data. Because it has
some lower latency requirements, a shortcut to efficiently get the
active stages from the active listener was added to the AppStateStore.

Now that all UI code goes through the status store to get its data,
the FsHistoryProvider can be cleaned up to only replay event logs
when needed - that is, when there is no pre-existing disk store for
the application.

As part of this change I also modified the streaming UI to read the needed
data from the store, which was missed in the previous patch that made
JobProgressListener redundant.

### CODE REVIEW

**Comment by vanzin** [comment]
For context:

* Project link: https://issues.apache.org/jira/browse/SPARK-18085
* Upcoming PRs that build on this code: https://github.com/vanzin/spark/pulls

File: core/src/main/scala/org/apache/spark/SparkStatusTracker.scala [hunk 1]
```diff
@@ -89,39 +83,33 @@ class SparkStatusTracker private[spark] (sc: SparkContext) {
...
       new SparkExecutorInfoImpl(
-        bmId.host,
-        bmId.port,
-        status.cacheSize,
-        executorIdToRunningTasks.getOrElse(bmId.executorId, 0)
-      )
-    }
+        host,
+        port,
+        exec.maxMemory,
```
Comment (ajbozarth) [inline core/src/main/scala/org/apache/spark/SparkStatusTracker.scala]: Why this change from cacheSize to maxMemory, are they synonymous?

File: core/src/main/scala/org/apache/spark/TestUtils.scala [hunk 2]
```diff
@@ -232,6 +232,30 @@ private[spark] object TestUtils {
...
+   * @param numExecutors the number of executors to wait at least
+   * @param timeout time to wait in milliseconds
+   */
+  private[spark] def waitUntilExecutorsUp(
+      sc: SparkContext,
+      numExecutors: Int,
+      timeout: Long): Unit = {
+    val finishTime = System.nanoTime() + TimeUnit.MILLISECONDS.toNanos(timeout)
+    while (System.nanoTime() < finishTime) {
+      if (sc.statusTracker.getExecutorInfos.length > numExecutors) {
```
Comment (ajbozarth) [inline core/src/main/scala/org/apache/spark/TestUtils.scala]: Based on your comment `at least` above, shouldn't this be `>=`?

File: core/src/main/scala/org/apache/spark/deploy/history/FsHistoryProvider.scala [hunk 3]
```diff
@@ -315,48 +315,43 @@ private[history] class FsHistoryProvider(conf: SparkConf, clock: Clock)
         (new InMemoryStore(), true)
     }
 
-    val listener = if (needReplay) {
-      val _listener = new AppStatusListener(kvstore, conf, false,
+    if (needReplay) {
+      val replayBus = new ReplayListenerBus()
```
Comment (ajbozarth) [inline core/src/main/scala/org/apache/spark/deploy/history/FsHistoryProvider.scala]: Doesn't this val already exist?

File: core/src/main/scala/org/apache/spark/status/AppStatusListener.scala [hunk 4]
```diff
@@ -563,7 +575,7 @@ private[spark] class AppStatusListener(
   /** Flush all live entities' data to the underlying store. */
   def flush(): Unit = {
     val now = System.nanoTime()
-    liveStages.values.foreach { stage =>
+    liveStages.values.asScala.foreach { stage =>
```
Comment (ajbozarth) [inline core/src/main/scala/org/apache/spark/status/AppStatusListener.scala]: Why add `asScala`?

**Review by ajbozarth — COMMENTED** [review]
A few comments but overall this lgtm

File: core/src/main/scala/org/apache/spark/status/AppStatusListener.scala (hunk 4 above)
Comment (vanzin) [reply core/src/main/scala/org/apache/spark/status/AppStatusListener.scala]: `liveStages` is not a Scala map anymore.

File: core/src/main/scala/org/apache/spark/deploy/history/FsHistoryProvider.scala (hunk 3 above)
Comment (vanzin) [reply core/src/main/scala/org/apache/spark/deploy/history/FsHistoryProvider.scala]: Removed stale one.

File: core/src/main/scala/org/apache/spark/TestUtils.scala (hunk 2 above)
Comment (vanzin) [reply core/src/main/scala/org/apache/spark/TestUtils.scala]: `getExecutorInfos` includes the driver...

File: core/src/main/scala/org/apache/spark/deploy/history/FsHistoryProvider.scala (hunk 3 above)
Comment (ajbozarth) [reply core/src/main/scala/org/apache/spark/deploy/history/FsHistoryProvider.scala]: It's still there in this PR though (line 297)

File: core/src/main/scala/org/apache/spark/status/AppStatusListener.scala (hunk 4 above)
Comment (ajbozarth) [reply core/src/main/scala/org/apache/spark/status/AppStatusListener.scala]: Ah I missed that that class change was scala -> java

File: core/src/main/scala/org/apache/spark/SparkStatusTracker.scala (hunk 1 above)
Comment (vanzin) [reply core/src/main/scala/org/apache/spark/SparkStatusTracker.scala]:
```
  /** Return the memory used by caching RDDs */
  def cacheSize: Long = onHeapCacheSize.getOrElse(0L) + offHeapCacheSize.getOrElse(0L)
```

So you're right, this isn't correct.

**Review by ajbozarth — APPROVED** [review]
LGTM

**Review by squito — APPROVED** [review]
lgtm

**Comment by gengliangwang** [comment]
lgtm

**Comment by dongjoon-hyun** [comment]
Retest this please.

**Review by dongjoon-hyun — COMMENTED** [review]
+1, LGTM, too.

**Comment by vanzin** [comment]
Merging to master.
